core/src/wsv.rs

```diff
@@ -124,12 +125,54 @@ impl<W: WorldTrait> WorldStateView<W> {
         }
     }
 
+    #[cfg(feature = "roles")]
+    /// Returns a set of permission tokens granted to this account as part of roles and separately.
+    pub fn account_permission_tokens(
+        &self,
+        account: &Account,
+    ) -> iroha_data_model::account::Permissions {
+        let mut tokens = account.permission_tokens.clone();
+        for role_id in &account.roles {
+            if let Some(role) = self.world.roles.get(role_id) {
+                let mut role_tokens = role.permissions.clone();
+                tokens.append(&mut role_tokens);
+            }
+        }
+        tokens
+    }
+
+    #[allow(clippy::unused_self)]
+    #[cfg(not(feature = "roles"))]
+    /// Returns a set of permission tokens granted to this account as part of roles and separately.
+    pub fn account_permission_tokens(
+        &self,
+        account: &Account,
+    ) -> iroha_data_model::account::Permissions {
+        account.permission_tokens.clone()
+    }
+
+    /// Apply instructions to the `WorldStateView<W>`.
+    ///
+    /// # Errors
+    /// Can fail if execution of instructions fail(should be fine after validation)
+    fn proceed(&self, tx: &VersionedValidTransaction) -> Result<()> {
+        let tx = tx.as_inner_v1();
+
+        for instruction in &tx.payload.instructions {
+            instruction
+                .clone()
+                .execute(tx.payload.account_id.clone(), self)?;
+        }
+        // XXX: Should it just return `()`?
+        Ok(())
+    }
+
     /// Apply `CommittedBlock` with changes in form of **Iroha Special Instructions** to `self`.
     #[iroha_futures::telemetry_future]
     #[iroha_logger::log(skip(self, block))]
     pub async fn apply(&self, block: VersionedCommittedBlock) {
         for tx in &block.as_inner_v1().transactions {
-            if let Err(error) = tx.proceed(self) {
+            if let Err(error) = self.proceed(tx) {
```

Comment:
I'm not sure about `log` and forget approach. 

While we're at it, is there any chance of propagating the error up the call chain?